### ISSUE TITLE
Reactive.jl update broke my Escher widgets

### ISSUE THREAD

**Issue by brandonkmiller** [shashi/Escher.jl#114]
I can no longer get my Escher apps to run correctly after Pkg.update().  There appears to be some issue since moving from Input() to Signal() or lift to map in Reactive.jl.  After the update I first started my app and got the deprecation warnings - however, they appear to be more than warnings because without correcting the deprecations my app wouldn't load at all.  When I changed lift->map and Input->Signal my app loads but the tabs are broken.  Both pages in the tab() widget are now displayed on the same page and switching between the tab buttons doesn't do anything.  Furthermore, my leftbutton doesn't work. Everything just hangs.  Here's were it ends:

EDIT

Pkg.checkout("Escher"); Pkg.build("Escher") fixed the brunt of the issue.  I'd say in a perfect world that it would be awesome if all the versions in METADATA could stay compatible, but this isn't a perfect world ;)  Now I'm getting
``` 
Failed to push!
    WARNING: both Patchwork and Base export "Text"; uses of it in module Main must be qualified
Signal{Escher.FillColor}(Escher.FillColor(RGB{U8}(0.204,0.204,0.204),Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList(Union{Escher.FillColor,Escher.PackedAcross{Escher.AxisCenter}}[Escher.FillColor(RGB{U8}(0.204,0.204,0.204),Escher.PackedAcross{Escher.AxisCenter}(Escher.AxisCenter(),Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList([Escher.PadContent(Escher.Side[],Escher.Length{:em}(1.0),Escher.Container(Escher.Flow{Escher.Horizontal}(Escher.Horizontal(),Escher.TileList([Escher.Image("assets/img/cateye.png","")]),false)))]),false))),Escher.FillColor(RGB{U8}(0.333,0.333,0.333),Escher.PackedAcross{Escher.AxisCenter}(Escher.AxisCenter(),Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList([Escher.Width("max",Escher.Length{:em}(50.0),Escher.ChanSend(:tabschannel,Escher.Selectable(Escher.Tabs(Escher.TileList([Escher.PadContent(Escher.Side{Escher.Horizontal}[Escher.Left(),Escher.Right()],Escher.Length{:em}(2.0),Escher.Container(Escher.FontColor(RGB{U8}(1.0,1.0,1.0),Escher.TileList([Escher.Flow{Escher.Horizontal}(Escher.Horizontal(),Escher.TileList([Escher.WithFontType(Escher.Monospace(),Escher.TileList([Escher.WithFontSize{Escher.Length{:em}}(Escher.Length{:em}(2.0),Escher.TileList(ASCIIString["Management"]))]))]),false)])))),Escher.PadContent(Escher.Side{Escher.Horizontal}[Es
....
to node
....
UndefVarError: withlastchild not defined
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/behavior.jl:186
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:73
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:92
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:290
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:266
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/embellishment.jl:175
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:92
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:290
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/embellishment.jl:175
 in anonymous at /home/dsb/.julia/v0.4/Escher/src/cli/serve.jl:123
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:111
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:88
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:256
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:133
 in run at /home/dsb/.julia/v0.4/Reactive/src/core.jl:190
 in run at /home/dsb/.julia/v0.4/Reactive/src/core.jl:181
 [inlined code] from /home/dsb/.julia/v0.4/Reactive/src/core.jl:228
 in anonymous at task.jl:447
```

**Comment by r2dbg** [comment]
Pkg.checkout("Escher"); Pkg.build("Escher") fixed the brunt of the issue.  I'd say in a perfect world that it would be awesome if all the versions in METADATA could stay compatible, but this isn't a perfect world ;)  Now I'm getting 
Failed to push!
    WARNING: both Patchwork and Base export "Text"; uses of it in module Main must be qualified
Signal{Escher.FillColor}(Escher.FillColor(RGB{U8}(0.204,0.204,0.204),Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList(Union{Escher.FillColor,Escher.PackedAcross{Escher.AxisCenter}}[Escher.FillColor(RGB{U8}(0.204,0.204,0.204),Escher.PackedAcross{Escher.AxisCenter}(Escher.AxisCenter(),Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList([Escher.PadContent(Escher.Side[],Escher.Length{:em}(1.0),Escher.Container(Escher.Flow{Escher.Horizontal}(Escher.Horizontal(),Escher.TileList([Escher.Image("assets/img/cateye.png","")]),false)))]),false))),Escher.FillColor(RGB{U8}(0.333,0.333,0.333),Escher.PackedAcross{Escher.AxisCenter}(Escher.AxisCenter(),Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList([Escher.Width("max",Escher.Length{:em}(50.0),Escher.ChanSend(:tabschannel,Escher.Selectable(Escher.Tabs(Escher.TileList([Escher.PadContent(Escher.Side{Escher.Horizontal}[Escher.Left(),Escher.Right()],Escher.Length{:em}(2.0),Escher.Container(Escher.FontColor(RGB{U8}(1.0,1.0,1.0),Escher.TileList([Escher.Flow{Escher.Horizontal}(Escher.Horizontal(),Escher.TileList([Escher.WithFontType(Escher.Monospace(),Escher.TileList([Escher.WithFontSize{Escher.Length{:em}}(Escher.Length{:em}(2.0),Escher.TileList(ASCIIString["Management"]))]))]),false)])))),Escher.PadContent(Escher.Side{Escher.Horizontal}[Es
....
to node
....
UndefVarError: withlastchild not defined
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/behavior.jl:186
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:73
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:92
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:290
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:266
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/embellishment.jl:175
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:92
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:290
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/embellishment.jl:175
 in anonymous at /home/dsb/.julia/v0.4/Escher/src/cli/serve.jl:123
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:111
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:88
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:256
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:133
 in run at /home/dsb/.julia/v0.4/Reactive/src/core.jl:190
 in run at /home/dsb/.julia/v0.4/Reactive/src/core.jl:181
 [inlined code] from /home/dsb/.julia/v0.4/Reactive/src/core.jl:228
 in anonymous at task.jl:447

**Comment by r2dbg** [comment]
I'm replicating the same issue running the Escher examples - specifically mc.jl:
in call at /home/dsb/.julia/v0.4/Compose/src/patchable.jl:21
Failed to push!
    WARNING: both Patchwork and Base export "Text"; uses of it in module Main must be qualified
Signal{Escher.PadContent}(Escher.PadContent(Escher.Side[],Escher.Length{:em}(2.0),Escher.Container(Escher.Flow{Escher.Vertical}(Escher.Vertical(),Escher.TileList(Union{Escher.Class,Escher.PackedAcross{Escher.AxisCenter},Rea

...

MethodError: `isfinished` has no method matching isfinished(::Compose.Patchable)
 in draw at /home/dsb/.julia/v0.4/Compose/src/container.jl:430
 in compose_render at /home/dsb/.julia/v0.4/Escher/src/basics/lazyload.jl:54
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:46
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:92
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:290
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:266
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/tile.jl:92
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:290
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:462
 in render at /home/dsb/.julia/v0.4/Escher/src/basics/layout.jl:471
 in anonymous at /home/dsb/.julia/v0.4/Escher/src/cli/serve.jl:123
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:111
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:88
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in anonymous at /home/dsb/.julia/v0.4/Reactive/src/operators.jl:256
 in do_action at /home/dsb/.julia/v0.4/Reactive/src/core.jl:135
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:130
 in send_value! at /home/dsb/.julia/v0.4/Reactive/src/core.jl:133
 in run at /home/dsb/.julia/v0.4/Reactive/src/core.jl:190
 in run at /home/dsb/.julia/v0.4/Reactive/src/core.jl:181
 [inlined code] from /home/dsb/.julia/v0.4/Reactive/src/core.jl:228
 in anonymous at task.jl:447

I feel like this may be a compatibility issue between libraries.  Is anyone else experiencing this?

**Comment by r2dbg** [comment]
I'm at a lose :/

I've literally scraped my whole lib and v0.4 package dirs and ran nothing but Pkg.add("Escher") which should install all compatible deps correctly, right?  I'm still unable to get an Escher example to run.

**Comment by shashi** [comment]
Are you using the master branch of Escher?

https://github.com/JuliaLang/METADATA.jl/pull/4456 fix the patchwork issue...

**Comment by r2dbg** [comment]
I've tried multiple branches.  I've even tried a  fresh .julia directory letting Escher install it's deps.  You just get the lift/map and Input/Signal deprecations and the page hangs.

```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.3 (2016-01-12 21:37 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-unknown-linux-gnu

julia> Pkg.status("Escher")
 - Escher                        0.2.1

julia> Pkg.status("Reactive")
 - Reactive                      0.3.0

julia> Pkg.status("Mux")
 - Mux                           0.2.0

julia> Pkg.status("Patchwork")
 - Patchwork                     0.1.9
```

When I upgrade to master - Pkg.checkout("Escher"); Pkg.build("Escher") I get the following output when attempting to run form.jl:
```
Listening on 0.0.0.0:5555...
UndefVarError: capture not defined
 in anonymous at /home/r2dbg/.julia/v0.4/Escher/src/cli/serve.jl:170
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/Mux.jl:15
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/Mux.jl:8
 in splitquery at /home/r2dbg/.julia/v0.4/Mux/src/basics.jl:28
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/Mux.jl:8
 in wcatch at /home/r2dbg/.julia/v0.4/Mux/src/websockets_integration.jl:12
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/Mux.jl:8
 in todict at /home/r2dbg/.julia/v0.4/Mux/src/basics.jl:21
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/Mux.jl:12 (repeats 2 times)
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/Mux.jl:8
 in anonymous at /home/r2dbg/.julia/v0.4/Mux/src/server.jl:38
 in handle at /home/r2dbg/.julia/v0.4/WebSockets/src/WebSockets.jl:382
 in on_message_complete at /home/r2dbg/.julia/v0.4/HttpServer/src/HttpServer.jl:393
 in on_message_complete at /home/r2dbg/.julia/v0.4/HttpServer/src/RequestParser.jl:104
```

**Comment by shashi** [comment]
Fixed these things on master. Please try it out. You will need Reactive 0.3.0.

**Comment by r2dbg** [comment]
I'm getting the same error as issue #115 now

**Comment by shashi** [comment]
Please try now. I'm going to tag a fix release soon. Also added in a simple test for this stuff. You need to install `ComposeDiff` with `Pkg.add("ComposeDiff")` now to use Gadfly or Compose.

**Comment by shashi** [comment]
Tagged a release

**Comment by r2dbg** [comment]
When trying examples on Julia 0.5.0 I get:

```
Listening on 0.0.0.0:5555...
WARNING: both Reactive and Base export "foreach"; uses of it in module Main must be qualified
Error handling websocket connection:
UndefVarError: foreach not defined
 in anonymous at /home/dsb/.julia/v0.5/Escher/src/cli/serve.jl:161
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/Mux.jl:15
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/Mux.jl:8
 in splitquery at /home/dsb/.julia/v0.5/Mux/src/basics.jl:28
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/Mux.jl:8
 in wcatch at /home/dsb/.julia/v0.5/Mux/src/websockets_integration.jl:12
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/Mux.jl:8
 in todict at /home/dsb/.julia/v0.5/Mux/src/basics.jl:21
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/Mux.jl:12 (repeats 2 times)
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/Mux.jl:8
 in anonymous at /home/dsb/.julia/v0.5/Mux/src/server.jl:38
 in handle at /home/dsb/.julia/v0.5/WebSockets/src/WebSockets.jl:382
 in on_message_complete at /home/dsb/.julia/v0.5/HttpServer/src/HttpServer.jl:393
 in on_message_complete at /home/dsb/.julia/v0.5/HttpServer/src/RequestParser.jl:104^Cfatal: error thrown and no exception handler available.
```
However, in Julia stable it looks like your new tag is mostly workable.  mc.jl loads, but about 70% of the time when you click the start button nothing happens.  I'm having the same issue in my apps with the following code:
```julia
btn = Signal(Any, leftbutton)
```
When triggering with trigger!(s, :submit, button()) buttons work great (form.jl) - but the Signal(Any, leftbutton) style doesn't seem to work right :disappointed: (or at all in my applications).

**Comment by r2dbg** [comment]
To provide some context here's how I'm trying to execute a form:

```julia
dataview = Signal((Any, status_vbox("Upload a data file"))
filename = Signal(Any, "")
protocol = Signal(0)
filetype = Signal(Any, "0")
btn      = Signal(Any, leftbutton)

result = map(sampleon(btn, filename),
    sampleon(btn, protocol),
    sampleon(btn, filetype);
    typ=Any, init=nothing) do f, p, t
    run_processor(f, p, parse(Int, t), running, dataview)
end
```

I use this abstraction so I can control a spinner to show busy operations when processing data. run_processor crunches a static file and pushes the results as a vbox() into dataview.  However, when I click the button nothing happens.  No spinner and no backend terminal debug output.

**Comment by dioptre** [comment]
https://github.com/shashi/Escher.jl/pull/117

I fixed the foreach but now I'm getting start reactive errors https://github.com/shashi/Escher.jl/issues/116

How about you @r2dbg ?

**Comment by randyzwitch** [comment]
`foreach` ambiguity resolved with #132

**Comment by brandonkmiller** [comment]
I just tested the examples a day ago (since #132 was resolved) and still mc.jl only fires on button click after multiple refreshes.  Yes - the foreach issue helped on Julia v0.5 but I am still not getting the examples to behave correctly.

As I stated before, when I run the mc.jl example and repeatedly adjust the parameters and submit (without refreshing the page), after a few times the animated graphic no longer renders.  Perhaps this is a separate issue.
